src/main/java/mods/railcraft/common/blocks/tracks/BlockTrack.java

```diff
@@ -231,17 +235,39 @@ public void onEntityCollidedWithBlock(World world, int x, int y, int z, Entity e
         if (!(track instanceof IElectricGrid)) return;
 
         IElectricGrid.ChargeHandler chargeHandler = ((IElectricGrid) track).getChargeHandler();
-        if (chargeHandler.getCharge() > 2000)
-            if (entity instanceof EntityPlayer && ItemOveralls.isPlayerWearing((EntityPlayer) entity)) {
-                if (!((EntityPlayer) entity).capabilities.isCreativeMode && MiscTools.RANDOM.nextInt(150) == 0) {
-                    EntityPlayer player = ((EntityPlayer) entity);
-                    ItemStack pants = player.getCurrentArmor(MiscTools.ArmorSlots.LEGS.ordinal());
-                    player.setCurrentItemOrArmor(
-                            MiscTools.ArmorSlots.LEGS.ordinal() + 1,
-                            InvTools.damageItem(pants, 1));
+        if (chargeHandler.getCharge() > 2000) {
+            if (entity instanceof EntityPlayer player) {
+                ItemStack pants = player.getCurrentArmor(MiscTools.ArmorSlots.LEGS.ordinal());
+                if (pants != null && pants.getItem() != null) {
+                    if (ModuleManager.isModuleLoaded(Module.GREGTECH)) {
+                        // The current GT Hazard API has no concept of "damage on use", so we'll exclude that here.
+                        // Can be updated if it's more important to have the pants damaged than behaving like a GT
+                        // hazard.
+                        if (HazardProtection.protectsAgainstHazard(pants, Hazard.ELECTRICAL)) {
+                            return;
+                        }
+                        ItemStack boots = player.getCurrentArmor(MiscTools.ArmorSlots.BOOTS.ordinal());
+                        if (boots != null && HazardProtection.protectsAgainstHazard(boots, Hazard.ELECTRICAL)) {
+                            return;
+                        }
+                    } else {
+                        // should we move up to modern type pattern syntax here for a newer Java version?
+                        if (pants.getItem() instanceof ISafetyPants safetyPants
+                                && safetyPants.blocksElectricTrackDamage(pants)) {
+                            if (!player.capabilities.isCreativeMode && MiscTools.RANDOM.nextInt(150) == 0) {
+                                safetyPants.onShock(pants, player);
+                            }
+                            return;
+                        }
+                    }
                 }
-            } else if (((EntityLivingBase) entity).attackEntityFrom(RailcraftDamageSource.TRACK_ELECTRIC, 2))
-                chargeHandler.removeCharge(2000);
+            }
+            tryZap((EntityLivingBase) entity, chargeHandler);
```

Comment:
I'm not sure why we still have a cast to `(EntityLivingBase) entity` here. It seems like if the entity was not living, we could've exited out of this method a long time ago. We should probably check and cast this properly through `instanceof` near the start of the method.